### PR TITLE
Add sudachi_projection option to BertJapaneseTokenizer

### PR DESCRIPTION
**Pull request by hiroshi-matsuda-rit**
# What does this PR do?

A new feature of SudachiPy v0.6.8 allows normalization of words based on Japanese morphological analysis.

https://github.com/WorksApplications/sudachi.rs/issues/230

This morphology-based normalization functionality, named "projection" in SudachiPy, makes Japanese sub-tokenization more efficient and can improve transformer performance.

Very few changes are required to add `sudachi_projection` option to `BertJapaneseTokenizer`, and the models that do not specify `sudachi_projection` option can be used as before in environments using older versions of SudachiPy.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

- text models: @ArthurZucker and @younesbelkada

### CODE REVIEW

**Comment by hiroshi-matsuda-rit** [comment]
Example model for sudachi_projection option:
https://huggingface.co/hiroshi-matsuda-rit/bert-base-sudachitra-v11/blob/main/tokenizer_config.json#L16

**Comment by ArthurZucker** [comment]
Nice 🔥 Feel free to ping me for a review!

**Comment by hiroshi-matsuda-rit** [comment]
@ArthurZucker All tests passed! Please review and merge this PR.

File: src/transformers/models/bert_japanese/tokenization_bert_japanese.py [hunk 1]
```diff
@@ -586,9 +589,15 @@ def __init__(
...
+        self.projection = sudachi_projection
+
+        sudachi_dictionary = dictionary.Dictionary(
             config_path=sudachi_config_path, resource_dir=sudachi_resource_dir, dict=sudachi_dict_type
-        ).create(self.split_mode)
+        )
+        if self.projection:
+            self.sudachi = sudachi_dictionary.create(self.split_mode, projection=self.projection)
+        else:
+            self.sudachi = sudachi_dictionary.create(self.split_mode)
```
Comment (ArthurZucker) [inline src/transformers/models/bert_japanese/tokenization_bert_japanese.py:600]:
```suggestion
         self.sudachi = sudachi_dictionary.create(self.split_mode, projection=self.projection)
```
this should work no? Unless call with none is not supported?

File: src/transformers/models/bert_japanese/tokenization_bert_japanese.py [hunk 2]
```diff
@@ -562,6 +563,8 @@ def __init__(
             **sudachi_resource_dir**: (*optional*) string
             **sudachi_dict_type**: (*optional*) string
                 dict type of sudachi, choose from "small", "core", "full".
+            **sudachi_projection**: (*optional*) string
+                Word projection mode of sudachi, choose from "surface", "normalized", "reading", "dictionary", "dictionary_and_surface", "normalized_and_surface", "normalized_nouns".
```
Comment (ArthurZucker) [inline src/transformers/models/bert_japanese/tokenization_bert_japanese.py:567]:
```suggestion
            **sudachi_projection**: (*optional*) string
                Word projection mode of sudachi, choose from `["surface", "normalized", "reading", "dictionary", "dictionary_and_surface", "normalized_and_surface", "normalized_nouns"]`.
```

**Review by ArthurZucker — COMMENTED** [review]
Looks good to me! I am glad that the test did not change! Thanks you for adding this new feature to the Sudachi tokenizer 

Let's add a small test where you pass `sudachi_kwargs` to a `BertJapaneseTokenizer` to make sure this is usable

File: src/transformers/models/bert_japanese/tokenization_bert_japanese.py (hunk 2 above)
Comment (hiroshi-matsuda-rit) [reply src/transformers/models/bert_japanese/tokenization_bert_japanese.py:567]: Looks nice!
I'd like to change above options for `sudachi_dict_type` in same way.

File: src/transformers/models/bert_japanese/tokenization_bert_japanese.py (hunk 1 above)
Comment (hiroshi-matsuda-rit) [reply src/transformers/models/bert_japanese/tokenization_bert_japanese.py:600]: This causes exceptions. Previous versions of SudachiPy do not have `projection` arg for `Dictionary.create()`.

Comment (hiroshi-matsuda-rit) [reply src/transformers/models/bert_japanese/tokenization_bert_japanese.py:600]: I'd like to add a comment to describe above comment.

**Review by ArthurZucker — COMMENTED** [review]
I think we need a sudachi version check for this feature then! Instead of if else ! Something similar to `version.parse(_torch_version) >= version.parse("2.1.1")` (from: https://github.com/huggingface/transformers/blob/64fa9f58e82216defb01740079f374609ab5ae17/src/transformers/utils/import_utils.py#L265-L277

with an exception if not supported

**Comment by hiroshi-matsuda-rit** [comment]
> Let's add a small test where you pass sudachi_kwargs to a BertJapaneseTokenizer to make sure this is usable

I added some test cases to pass kwargs to word tokenizers.

**Comment by hiroshi-matsuda-rit** [comment]
> I think we need a sudachi version check for this feature then!

Thanks for your suggestion. Let me explain the background.

I've been developing a Japanese Dependency parser GiNZA utilizing spacy-transformers and sudachipy, and its [ELECTRA model](https://huggingface.co/megagonlabs/transformers-ud-japanese-electra-base-ginza-520/tree/main) has become widely used for Japanese text analysis in recent years.
I'm wondering that certain users of GiNZA ELECTRA model may hope to keep sudachipy as is but they need to upgrade transformers to use newer models together with GiNZA.

This is the reason why I wrote the if else clause.

**Comment by hiroshi-matsuda-rit** [comment]
@ArthurZucker How should I do for this type of errors? It seems I do not have a permission to rerun the failed actions.

> FAILED tests/test_pipeline_mixin.py::ImageSegmentationPipelineTests::test_small_model_pt - requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))

**Comment by ArthurZucker** [comment]
Yes if else is alright don't worry! Just wanted to make sure someone that tries to use the new feature with an older version would get an error saying he needs to update the sudachi version to use it!

File: src/transformers/models/bert_japanese/tokenization_bert_japanese.py [hunk 3]
```diff
@@ -586,9 +589,12 @@ def __init__(
...
             raise ValueError("Invalid sudachi_split_mode is specified.")
 
-        self.sudachi = dictionary.Dictionary(
+        self.projection = sudachi_projection
+
+        sudachi_dictionary = dictionary.Dictionary(
             config_path=sudachi_config_path, resource_dir=sudachi_resource_dir, dict=sudachi_dict_type
-        ).create(self.split_mode)
+        )
+        self.sudachi = sudachi_dictionary.create(self.split_mode, projection=self.projection)
```
Comment (ArthurZucker) [inline src/transformers/models/bert_japanese/tokenization_bert_japanese.py:597]: as you mentioned this is not backward compatible as this was introduced later on in sudachi!

File: tests/models/bert_japanese/test_tokenization_bert_japanese.py
Comment (ArthurZucker) [inline tests/models/bert_japanese/test_tokenization_bert_japanese.py:1]: nice tests thanks!

**Review by ArthurZucker — COMMENTED** [review]
LGTM, we need to make sure we test the sudachi version an warn users that would set a projection that is not supported!

**Comment by hiroshi-matsuda-rit** [comment]
I implemented `is_sudachi_projection_available()` and `@require_sudachi_projection` to check the sudachipy version.
Does this additional implementation meet your requirements?

**Comment by ArthurZucker** [comment]
Reviewing!

File: setup.py [hunk 4]
```diff
@@ -164,7 +164,7 @@
     "sentencepiece>=0.1.91,!=0.1.92",
     "sigopt",
     "starlette",
-    "sudachipy>=0.6.6",
+    "sudachipy>=0.6.8",
```
Comment (ArthurZucker) [inline setup.py:167]: we usually change this only for security reasons

File: src/transformers/models/bert_japanese/tokenization_bert_japanese.py (hunk 3 above)
Comment (ArthurZucker) [reply src/transformers/models/bert_japanese/tokenization_bert_japanese.py:597]: I would use `if is_sudachi_projection_available` here, and else use the backward compatible beahviour!

**Review by ArthurZucker — COMMENTED** [review]
The test values should not be updated, but nice new test that was added!

**Comment by hiroshi-matsuda-rit** [comment]
@ArthurZucker Thanks! I just revised the changes as per your instructions.

File: src/transformers/models/bert_japanese/tokenization_bert_japanese.py [hunk 5]
```diff
@@ -586,9 +589,18 @@ def __init__(
...
+        sudachi_dictionary = dictionary.Dictionary(
             config_path=sudachi_config_path, resource_dir=sudachi_resource_dir, dict=sudachi_dict_type
-        ).create(self.split_mode)
+        )
+        if is_sudachi_projection_available():
+            self.sudachi = sudachi_dictionary.create(self.split_mode, projection=self.projection)
+        else:
+            assert (
+                self.projection is None
+            ), "You need to install sudachipy>=0.6.8 to specify `projection` field in sudachi_kwargs."
```
Comment (ArthurZucker) [inline src/transformers/models/bert_japanese/tokenization_bert_japanese.py:602]:
```suggestion
            Raise ImportError("You need to install sudachipy>=0.6.8 to specify `projection` field in sudachi_kwargs."
```

**Review by ArthurZucker — COMMENTED** [review]
That should be the last nit!

**Comment by hiroshi-matsuda-rit** [comment]
@ArthurZucker Thank you. I have completed the fix.

**Review by ArthurZucker — APPROVED** [review]
LGTM! Thank you for iterating and protecting the imports!

**Pull request merged**